secret_store/src/key_server_cluster/cluster_connections_net.rs

```diff
@@ -0,0 +1,543 @@
+			node_id: connection.node_id,
+			node_address: connection.address,
+			is_inbound: is_inbound,
+			stream: connection.stream,
+			key: connection.key,
+			last_message_time: RwLock::new(Instant::now()),
+		}
+	}
+
+	/// Get last message time.
+	pub fn last_message_time(&self) -> Instant {
+		*self.last_message_time.read()
+	}
+
+	/// Update last message time
+	pub fn set_last_message_time(&self, last_message_time: Instant) {
+		*self.last_message_time.write() = last_message_time
+	}
+
+	/// Returns future that sends encrypted message over this connection.
+	pub fn send_message_future(&self, message: Message) -> WriteMessage<SharedTcpStream> {
+		write_encrypted_message(self.stream.clone(), &self.key, message)
+	}
+
+	/// Returns future that reads encrypted message from this connection.
+	pub fn read_message_future(&self) -> ReadMessage<SharedTcpStream> {
+		read_encrypted_message(self.stream.clone(), self.key.clone())
+	}
+}
+
+impl Connection for NetConnection {
+	fn is_inbound(&self) -> bool {
+		self.is_inbound
+	}
+
+	fn node_id(&self) -> &NodeId {
+		&self.node_id
+	}
+
+	fn node_address(&self) -> String {
+		format!("{}", self.node_address)
+	}
+
+	fn send_message(&self, message: Message) {
+		execute(&self.executor, self.send_message_future(message).then(|_| Ok(())));
+	}
+}
+
+impl NetConnectionsData {
+	/// Executes closure for each active connection.
+	pub fn active_connections(&self) -> Vec<Arc<NetConnection>> {
+		self.container.read().connections.values().cloned().collect()
+	}
+
+	/// Executes closure for each disconnected node.
+	pub fn disconnected_nodes(&self) -> Vec<(NodeId, SocketAddr)> {
+		let container = self.container.read();
+		container.nodes.iter()
+			.filter(|(node_id, _)| !container.connections.contains_key(node_id))
+			.map(|(node_id, addr)| (node_id.clone(), addr.clone()))
+			.collect()
+	}
+
+	/// Try to insert new connection. Returns true if connection has been inserted.
+	/// Returns false (and ignores connections) if:
+	/// - we do not expect connection from this node
+	/// - we are already connected to the node and existing connection 'supersede'
+	///   new connection by agreement
+	pub fn insert(&self, connection: Arc<NetConnection>) -> bool {
+		let mut container = self.container.write();
+		if !container.nodes.contains_key(connection.node_id()) {
+			trace!(target: "secretstore_net", "{}: ignoring unknown connection from {} at {}",
+				self.self_key_pair.public(), connection.node_id(), connection.node_address());
+			return false;
+		}
+
+		if container.connections.contains_key(connection.node_id()) {
+			// we have already connected to the same node
+			// the agreement is that node with lower id must establish connection to node with higher id
+			if (self.self_key_pair.public() < connection.node_id() && connection.is_inbound())
+				|| (self.self_key_pair.public() > connection.node_id() && !connection.is_inbound()) {
+				return false;
+			}
+		}
+
+		let node = connection.node_id().clone();
```

Comment:
connection.node_id() called several times, you could call it once and then pass the node the functions/methods where it is used! But, perhaps the compiler may handle this for you :P 

```suggestion
		let node = *connection.node_id();
```